In [1]:
import pandas as pd

In [3]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error, r2_score

In [4]:
df = pd.read_csv('car_purchasing.csv', encoding='latin1')
#drop the column
df = df.drop(columns=['customer name', 'customer e-mail'])

#clean the data
df = pd.get_dummies(df, columns=['country'], drop_first=True)

X = df.drop('car purchase amount', axis=1)
y = df['car purchase amount']

In [5]:
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# Train model
model = RandomForestRegressor(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

# Predict and evaluate
y_pred = model.predict(X_test)
rmse = mean_squared_error(y_test, y_pred, squared=False)
r2 = r2_score(y_test, y_pred)

print("Root Mean Squared Error (RMSE):", rmse)
print("R² Score:", r2)

Root Mean Squared Error (RMSE): 2385.5953608542773
R² Score: 0.9472918524259603


RMSE ~ 2385.60: This is low relative to target (car purchase amount), indicating the model makes small errors on average.
R² Score ~ 0.95: The model explains ~95% of the variance in car purchase amounts.

In [7]:
import matplotlib.pyplot as plt

In [10]:
import joblib
joblib.dump(model, 'car_purchase_predictor.pkl')

['car_purchase_predictor.pkl']

In [21]:
# Step 1: Get the list of columns in the training data
train_columns = X_train_df.columns

# Step 2: Ensure the 'country' column exists in the new data, and strip any whitespace
new_data_df.columns = new_data_df.columns.str.strip()  # Remove leading/trailing spaces

# Check if 'country' exists in the new data
if 'country' in new_data_df.columns:
    # Step 3: One-hot encode 'country' in both datasets
    new_data_df_encoded = pd.get_dummies(new_data_df, columns=['country'])
    X_train_encoded = pd.get_dummies(X_train_df, columns=['country'])
    
    # Step 4: Align the columns of the new data to match the training data
    new_data_df_encoded = new_data_df_encoded.reindex(columns=X_train_encoded.columns, fill_value=0)

    # Step 5: Scale the new data using the same scaler
    new_data_scaled = scaler.transform(new_data_df_encoded)

    # Step 6: Make the prediction
    prediction = model.predict(new_data_scaled)

    # Step 7: Print the predicted amount
    print(f"Predicted Car Purchase Amount: ${prediction[0]:,.2f}")
else:
    print("Error: 'country' column not found in the new data.")






Error: 'country' column not found in the new data.


In [22]:
print(new_data_df.columns)


Index(['age', 'income', 'credit_score', 'loan_amount', 'home_value',
       'country_0', 'country_1', 'country_2', 'country_3', 'country_4',
       ...
       'country_200', 'country_201', 'country_202', 'country_203',
       'country_204', 'country_205', 'country_206', 'country_207',
       'country_208', 'country_209'],
      dtype='object', length=215)


In [23]:
# Step 1: One-hot encode the country column for the training data (if not already done)
X_train_encoded = pd.get_dummies(X_train_df, columns=['country'])

# Step 2: One-hot encode the 'country' column for the new data (if not already done)
new_data_df_encoded = pd.get_dummies(new_data_df, columns=['country'])

# Step 3: Reindex the new data to match the training data columns
# This will add missing columns and fill them with 0s
new_data_df_encoded = new_data_df_encoded.reindex(columns=X_train_encoded.columns, fill_value=0)

# Step 4: Scale the new data using the same scaler
new_data_scaled = scaler.transform(new_data_df_encoded)

# Step 5: Make the prediction
prediction = model.predict(new_data_scaled)

# Step 6: Print the predicted amount
print(f"Predicted Car Purchase Amount: ${prediction[0]:,.2f}")


KeyError: "None of [Index(['country'], dtype='object')] are in the [columns]"

In [24]:
# Step 1: One-hot encode the 'country' column for the training data (if not already done)
X_train_encoded = pd.get_dummies(X_train_df, columns=['country'])

# Step 2: Ensure the new data already has one-hot encoded 'country' columns (skip encoding step for new data)
# Step 3: Reindex the new data to match the training data columns
new_data_df_encoded = new_data_df.reindex(columns=X_train_encoded.columns, fill_value=0)

# Step 4: Scale the new data using the same scaler
new_data_scaled = scaler.transform(new_data_df_encoded)

# Step 5: Make the prediction
prediction = model.predict(new_data_scaled)

# Step 6: Print the predicted amount
print(f"Predicted Car Purchase Amount: ${prediction[0]:,.2f}")



KeyError: "None of [Index(['country'], dtype='object')] are in the [columns]"